File: cslox/cslox/Scanning/Scanner.cs
namespace cslox.Scanning;

using static TokenType;

class Scanner
{
	private readonly string source;
	private List<Token> tokens { get; } = new List<Token>();
	private int start = 0;
	private int current = 0;
	private int line = 1;

	private static Dictionary<string, TokenType> Keywords = new Dictionary<string, TokenType>
	{
		["and"] = AND,
		["class"] = CLASS,
		["else"] = ELSE,
		["false"] = FALSE,
		["for"] = FOR,
		["fun"] = FUN,
		["if"] = IF,
		["nil"] = NIL,
		["or"] = OR,
		["print"] = PRINT,
		["return"] = RETURN,
		["super"] = SUPER,
		["this"] = THIS,
		["true"] = TRUE,
		["var"] = VAR,
		["while"] = WHILE,
	};

	public Scanner(string source)
	{
		this.source = source;
	}

	private bool IsAtEnd() => current >= source.Length;
	private char Advance() => source[current++];

	public List<Token> ScanTokens()
	{
		while (!IsAtEnd())
		{
			start = current;
			ScanToken();
		}

		tokens.Add(new Token(EOF, "", null, line));
		return tokens;
	}

	private void AddToken(TokenType type) => AddToken(type, null);

	private void AddToken(TokenType type, object? literal)
	{
		var text = source.Substring(start, current - start);
		tokens.Add(new Token(type, text, literal, line));
	}

	private bool Match(char expected)
	{
		if (IsAtEnd())
		{
			return false;
		}

		if (source[current] != expected)
		{
			return false;
		}

		current++;
		return true;
	}

	private char Peek()
	{
		if (IsAtEnd())
		{
			return '\0';
		}

		return source[current];
	}

	private char PeekNext()
	{
		if (current + 1 > source.Length)
		{
			return '\0';
		}
		return source[current + 1];
	}

	private void ScanString()
	{
		while (Peek() != '"' && !IsAtEnd())
		{
			if (Peek() == '\n')
			{
				line++;
			}
			Advance();
		}

		if (IsAtEnd())
		{
			Lox.Error(line, "Unterminated string.");
			return;
		}

		// The closing ".
		Advance();

		// Trim the surrounding quotes.
		string value = source.Substring(start + 1, current - start - 1);
		AddToken(STRING, value);
	}

	private bool IsDigit(char c) => (c >= '0' && c <= '9');

	private void ScanNumber()
	{
		while (IsDigit(Peek()))
		{
			Advance();
		}

		if (Peek() == '.' && IsDigit(PeekNext()))
		{
			// Consume the "."
			Advance();

			while (IsDigit(Peek()))
			{
				Advance();
			}
		}

		AddToken(NUMBER, Double.Parse(source.Substring(start, current - start)));
	}

	private bool IsAlpha(char c) =>
		(c >= 'a' && c <= 'z') ||
		(c >= 'A' && c <= 'Z') ||
		c == '_';

	private bool IsAlphaNumeric(char c) => IsAlpha(c) || IsDigit(c);

	private void ScanIdentifier()
	{
		while (IsAlphaNumeric(Peek()))
		{
			Advance();
		}

		var text = source.Substring(start, current - start);
		var type = IDENTIFIER;
		if (Keywords.ContainsKey(text))
		{
			type = Keywords[text];
		}

		AddToken(type);
	}

	private void ScanBlockComment()
	{
		var depth = 0;

        while (!IsAtEnd())
		{
            var c = Advance();
            switch (c)
            {
                case '/':
                    if (Match('*'))
                    {
                        depth++;
                    }
                    break;
                case '*':
                    if (Match('/'))
                    {
                        if (depth == 0)
                        {
                            return;
                        }
                        depth--;
                    }
                    break;
                case '\n':
                    line++;
                    break;
            }
        }
		Lox.Error(line, "Unclosed block comment");
    }

    private void ScanToken()
	{
		char c = Advance();
		switch (c)
		{
			case '(': AddToken(LEFT_PAREN); break;
			case ')': AddToken(RIGHT_PAREN); break;
			case '{': AddToken(LEFT_BRACE); break;
			case '}': AddToken(RIGHT_BRACE); break;
			case ',': AddToken(COMMA); break;
			case '.': AddToken(DOT); break;
			case '-': AddToken(MINUS); break;
			case '+': AddToken(PLUS); break;
			case ';': AddToken(SEMICOLON); break;
			case '*': AddToken(STAR); break;
			case '!':
				AddToken(Match('=') ? BANG_EQUAL : BANG);
				break;
			case '=':
				AddToken(Match('=') ? EQUAL_EQUAL : EQUAL);
				break;
			case '<':
				AddToken(Match('=') ? LESS_EQUAL : LESS);
				break;
			case '>':
				AddToken(Match('=') ? GREATER_EQUAL : GREATER);
				break;
			case '/':
				if (Match('/'))
				{
					// A comment goes until the end of the line.
					while (Peek() != '\n' && !IsAtEnd())
					{
						Advance();
					}
                }
				else if (Match('*'))
				{
					ScanBlockComment();
				}
				else
				{
					AddToken(SLASH);
				}
				break;
			case ' ':
			case '\t':
			case '\r':
				// Ignore whitespace
				break;
			case '\n':
				line++;
				break;
			case '"': ScanString(); break;

            default:
				if (IsDigit(c))
				{
					ScanNumber();
				}
				else if (IsAlpha(c))
				{
					ScanIdentifier();
				}
				else
				{
					Lox.Error(line, "Unexpected character.");
				}
				break;
		}
	}
}

File: tree-walk-interpreter/go/internal/scanner/scanner.go
package scanner

import (
	"strconv"

	loxError "github.com/kashifsoofi/go-lox/internal/error"
)

var (
	source  []rune
	tokens  []*Token
	start   int
	current int
	line    int
)

var keywordsTokenTypeMap = map[string]TokenType{
	"and":    And,
	"class":  Class,
	"else":   Else,
	"false":  False,
	"for":    For,
	"fun":    Fun,
	"if":     If,
	"nil":    Nil,
	"or":     Or,
	"print":  Print,
	"return": Return,
	"super":  Super,
	"this":   This,
	"true":   True,
	"var":    Var,
	"while":  While,
}

func ScanTokens(src string) []*Token {
	source = []rune(src)
	tokens = make([]*Token, 0)
	start = 0
	current = 0
	line = 1

	scanTokens()

	return tokens
}

func scanTokens() {
	for !isAtEnd() {
		// We are at the start of the next lexeme.
		start = current
		scanToken()
	}

	tokens = append(tokens, NewToken(EOF, "", nil, line))
}

func isAtEnd() bool {
	return current >= len(source)
}

func advance() rune {
	r := source[current]
	current++
	return r
}

func peek() rune {
	if isAtEnd() {
		return rune(0)
	}

	return source[current]
}

func peekNext() rune {
	if current+1 >= len(source) {
		return rune(0)
	}

	return source[current+1]
}

func match(r rune) bool {
	if isAtEnd() {
		return false
	}

	if source[current] != r {
		return false
	}

	current++
	return true
}

func isDigit(r rune) bool {
	return r >= '0' && r <= '9'
}

func isAlpha(r rune) bool {
	return (r >= 'a' && r <= 'z') ||
		(r >= 'A' && r <= 'Z') ||
		r == '_'
}

func isAlphaNumeric(r rune) bool {
	return isAlpha(r) || isDigit(r)
}

func scanToken() {
	r := advance()
	switch r {
	case '(':
		addToken(LeftParen)
	case ')':
		addToken(RightParen)
	case '{':
		addToken(LeftBrace)
	case '}':
		addToken(RightBrace)
	case ',':
		addToken(Comma)
	case '.':
		addToken(Dot)
	case '-':
		addToken(Minus)
	case '+':
		addToken(Plus)
	case ';':
		addToken(Semicolon)
	case '*':
		addToken(Star)
	case '!':
		if match('=') {
			addToken(BangEqual)
		} else {
			addToken(Bang)
		}
	case '=':
		if match('=') {
			addToken(EqualEqual)
		} else {
			addToken(Equal)
		}
	case '<':
		if match('=') {
			addToken(LessEqual)
		} else {
			addToken(Less)
		}
	case '>':
		if match('=') {
			addToken(GreaterEqual)
		} else {
			addToken(Greater)
		}
	case '/':
		if match('/') {
			for peek() != '\n' && !isAtEnd() {
				advance()
			}
		} else {
			addToken(Slash)
		}
	case ' ':
	case '\r':
	case '\t':
		break
	case '\n':
		line++
	case '"':
		scanString()
	default:
		if isDigit(r) {
			scanNumber()
		} else if isAlpha(r) {
			scanIdentifier()
		} else {
			loxError.Error(line, "Unexpected character.")
		}
	}
}

func scanString() {
	for peek() != '"' && !isAtEnd() {
		if peek() == '\n' {
			line++
		}
		advance()
	}

	if isAtEnd() {
		loxError.Error(line, "Unterminated string.")
		return
	}

	// The closing ".
	advance()

	value := source[start+1 : current-1]
	addTokenWithLiteral(String, string(value))
}

func scanNumber() {
	for isDigit(peek()) {
		advance()
	}

	if peek() == '.' && isDigit(peekNext()) {
		// Consume the "."
		advance()

		for isDigit(peek()) {
			advance()
		}
	}

	s := string(source[start:current])
	n, _ := strconv.ParseFloat(s, 64)
	addTokenWithLiteral(Number, n)
}

func scanIdentifier() {
	for isAlphaNumeric(peek()) {
		advance()
	}

	text := string(source[start:current])
	tokenType, ok := keywordsTokenTypeMap[text]
	if !ok {
		tokenType = Identifier
	}

	addToken(tokenType)
}

func addToken(tokenType TokenType) {
	addTokenWithLiteral(tokenType, nil)
}

func addTokenWithLiteral(tokenType TokenType, literal interface{}) {
	text := string(source[start:current])
	tokens = append(tokens, NewToken(tokenType, text, literal, line))
}
